<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Chris Talbot - Honors Thesis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Chris Talbot</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./CV.html"> 
<span class="menu-text">CV</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./honors-thesis.html" aria-current="page"> 
<span class="menu-text">Thesis</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools tools-wide">
    <a href="https://www.github.com/chris-a-talbot" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
    <a href="https://twitter.com/chris__talbot" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <a href="https://www.linkedin.com/in/chris-talbot" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-linkedin"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">On this page</h2>
   
  <ul class="collapse">
  <li><a href="#open-research-statement" id="toc-open-research-statement" class="nav-link active" data-scroll-target="#open-research-statement">Open Research Statement</a></li>
  <li><a href="#keywords" id="toc-keywords" class="nav-link" data-scroll-target="#keywords">Keywords</a></li>
  <li><a href="#abstract" id="toc-abstract" class="nav-link" data-scroll-target="#abstract">Abstract</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#methods" id="toc-methods" class="nav-link" data-scroll-target="#methods">Methods</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results">Results</a></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion">Discussion</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#author-contributions" id="toc-author-contributions" class="nav-link" data-scroll-target="#author-contributions">Author Contributions</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
<div class="quarto-other-links"><h2>Other Links</h2><ul><li><a href="files/honors_thesis.pdf"><i class="bi bi-link-45deg"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Honors Thesis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>**This honors thesis was submit to the University of Michigan for honors recognition, but has not been peer-reviewed and is not considered publication-ready.</p>
<p><span class="big"><strong>Patterns of floral color in communities of common Northeast American wildflowers</strong></span></p>
<p><strong>Chris Talbot<sup>1</sup>, Marjorie Weber<sup>1</sup></strong></p>
<p><sup><em>1</em></sup><em>Department of Ecology and Evolutionary Biology, University of Michigan, 1105 North University Avenue, Ann Arbor, MI 48109-1085, USA.</em></p>
<section id="open-research-statement" class="level2">
<h2 class="anchored" data-anchor-id="open-research-statement">Open Research Statement</h2>
<p>All data were retrieved from public sources, which are cited in-text. All novel code is permanently stored at <a href="https://github.com/chris-a-talbot/flower_color_diversity" class="uri">https://github.com/chris-a-talbot/flower_color_diversity</a>.</p>
</section>
<section id="keywords" class="level2">
<h2 class="anchored" data-anchor-id="keywords">Keywords</h2>
<p>biodiversity, citizen science, phenology, plant-pollinator interactions, sympatry</p>
</section>
<section id="abstract" class="level2">
<h2 class="anchored" data-anchor-id="abstract">Abstract</h2>
<p>Floral color is a captivating, complex trait, displaying striking variation across species and plant communities. A range of factors, including biotic interactions, abiotic conditions, and phylogenetic relationships is hypothesized to contribute to emergent patterns of floral color across plant communities, potentially generating overdispersed, underdispersed, or random patterns of floral color within communities. However, the generalizability of observed floral color patterns across communities remains uninvestigated for many floras. Here, we zoom in on the North American Eastern Wildflowers to ask: 1) Are species with similar colors more or less likely to co-occur in time and/or space? 2) Does the relationship between floral color and co-occurrence differ based on species relatedness? We assembled a large dataset of 1,096 species of North American flowering plants, and gathered data on floral color, phenology, distribution, and phylogenetic relationships. We selected 100 well-sampled 10 square kilometer communities and calculated metrics of community evenness in floral color categories and community phylogenetic dispersion. We found that overall, across North American Eastern wildflower communities, wildflowers are significantly more similar in color than expected by chance. However, contrary to hypotheses, the degree of over/underdispersion varied across communities, and was not not correlated with the phylogenetic over- or underdispersion of species in the community. Together, these results indicate that at the broad scale, patterns of color in wildflower communities of Northeast America are consistent with scenarios in which factors such as phylogenetic convergence filtering, and dominant pollinator preference shape emergent patterns of floral color across the landscape.</p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Flowering plants display a striking amount of variation in floral color, ranging from inconspicuous greens and browns, to eye-catching reds, blues, and yellows. The diversity in floral color exhibited across the globe has led to its use as a study system for evolutionary and ecological questions, including studies of adaptation, speciation, and evolutionary developmental biology (e.g., Clegg &amp; Durbin, 2000; Sobel &amp; Stresfield, 2013; Erickson &amp; Pessoa, 2022). However, despite the popularity of floral color in public and scientific arenas, the patterns of color composition across large-scale flowering plant communities remains understudied for many floras. Here, we examine the spatial distribution of broad floral colors in Northeast American flowering plants.</p>
<p>Several hypotheses attempt to explain the distribution of floral colors across a landscape. On the one hand, we may expect biotic factors to filter or select for highly divergent floral communities, whereby species in a given community have more dissimilar floral colors than expected by chance. Pollinators in particular are known to exhibit selective pressure for divergent floral colors. Innate and learned pollinator color preferences generate divergent pollinator-mediated selection on floral color and may result in overdispersion for floral color in communities (Weiss &amp; Papaj, 2003; Reverté et al., 2016; Muñoz-Galicia et al., 2021; Trunschke et al., 2021). Pollinator constancy, the preference of pollinators to repeatedly select flowers of a select few species or morphs, may contribute to divergent floral color within communities by incentivizing unique floral colors that attract pollinator specialists (Grossenbacher &amp; Whittall, 2014; Erickson &amp; Pessoa, 2022).</p>
<p>Alternatively, rather than divergence, several hypotheses suggest that floral color at the community scale should display convergence or underdispersion, whereby species in the same place tend to share the same color more than expected by chance. Both biotic and abiotic factors could generate similar floral color within a community. First, species within a community are subject to similar abiotic conditions, including temperature, UV irradiance, and precipitation. As such, if abiotic factors play a significant role in shaping floral color, we might expect all flowers within a community to display similar floral colors (Willmer, 2011). For example, darker floral colors have been associated with colder regions, as their dark color may increase warmth within petals, thereby increasing pollinator attraction (Ahmad et al., 2022). Furthermore, Koski and Ashman (2016) found a strong correlation between UV pigmentation and environmental UV irradiance for flowers in the Potentilleae tribe (Rosaceae). Variations in floral pigmentation are also associated with resistance to extreme environments, particularly drought resistance (Warren &amp; Mackenzie, 2001). At a given point in time, communities tend to have a small number of dominant pollinators. If the benefits of appealing to the color preferences of the dominant pollinator can outweigh the effects of any diverging processes, flowers in a community may tend towards the same color – specifically, the preferred color of the dominant pollinator during their blooming period (Kemp et al., 2019; Bergamo et al., 2020). Similarly, if pollinators prefer particular colors, pollinator constancy may counterintuitively incentivize convergent floral color evolution such that species can tap into the pollinators of other, similar species (Erickson &amp; Pessoa, 2022).</p>
<p>Finally, emergent patterns of floral color across communities may be random rather than over- or under-dispersed. Random emergent patterns at the community scale could occur if multiple factors hypothesized to shape the distribution of floral color occur concurrently. Additionally, although the effects of random and neutral processes on floral color require further investigation, it is not unreasonable to suspect that processes such as genetic drift and gene flow may play a noteworthy role in determining floral color (Sapir et al., 2021), which could also contribute to random patterns at the community scale.</p>
<p>When considering patterns of floral color dispersion in communities, it is also crucial to investigate evolutionary relationships (the phylogenetic dispersion) of species therewithin. Processes that may impact over- or under-dispersion of floral color phenotypes are hypothesized to have varying effects based on the phylogenetic structure of a community. For instance, divergent floral color may prove particularly beneficial to closely related species in a community, as divergence may facilitate increased coexistence for closely related species by avoiding deleterious hybridization and reproductive interference, allowing species to sustainably co-occur in space and time despite otherwise occupying similar areas of niche space (Grossenbacher &amp; Whitall, 2011). Alternatively, sharing floral color to attract a dominant pollinator may benefit distantly related species pairs but not closer relatives. Close relatives may also be expected to share floral color simply due to recent common ancestry (phylogenetic niche conservatism). Other floral traits, such as phenology, have been shown to exhibit high levels of trait conservatism, with congeners tending towards similar trait values (Davies et al., 2013). Thus, we may expect patterns of dispersion of floral colors in communities to shift with the relatedness of plants within that community.</p>
<p>While studies have evaluated patterns of floral color diversity within populations or small communities, to our knowledge, there is limited research on the generalizability of hypothesized patterns of floral color dispersion on a larger, landscape scale. Here, we ask: 1) are wildflowers with similar colors more or less likely to co-flower in space across Northeastern America? And 2) Do patterns of floral color dispersion differ based on species relatedness in communities? To address our questions, we gathered data on 1,096 species of wildflower found across North America. Our dataset includes information on floral color, phenology, range, and range overlap, as well as a phylogeny of our species. We used these data to investigate floral color patterns and relationships between floral color and relatedness.</p>
</section>
<section id="methods" class="level2">
<h2 class="anchored" data-anchor-id="methods">Methods</h2>
<p><em>Species Selection &amp; Phenotyping</em></p>
<p>We phenotyped the floral color of species using Newcomb’s Wildflower Guide, a field guide to wildflowers in Northeast America (hereafter “the field guide”) (Newcomb, 1977). We selected this guide for its categorization of each species into one or more discrete color categories. We standardized species names to the binomial nomenclature used by the Leipzig Catalogue of Vascular Plants (LCVP) using the LCVP and lcvplants packages in R (Freiberg et al., 2020). Standardization of nomenclature created some duplicates in our species list due to the reclassification of subspecies; we removed duplicate species from the data. We also standardized species families using the LCVP.&nbsp;</p>
<p>We identified the nine most common and broad color categories in the field guide - red, orange, yellow, green, blue, purple, pink, white, and brown. We placed each species into one or more of these color categories based on their descriptions in the field guide. Species labeled as red, scarlet, crimson, and maroon are categorized as “red;” purple, lavender, violet, lilac, and indigo species are categorized as “purple;” pink, magenta, and rose species are categorized as “pink;” white and cream species are categorized as “white.”&nbsp;</p>
<p><img src="images/categorization.png" class="img-fluid"></p>
<p><u><em>Figure 1:</em></u> <em>Species were collected from Newcomb’s Wildflower Guide (Newcomb, 1977); species names were standardized to Leipzig’s Catalogue of Vascular Plants (Freiberg et al., 2020); and species were placed into one or more of nine color categories in R. Each species is stored with binary variables for each color category, with 1 indicating the presence of a given color in the flowers of the species. Three examples are shown, with standardized species name, color description from the field guide, and the format of the consequent data.</em></p>
<p><em>Species Occurrences</em></p>
<p>Occurrence data for each species was collected using the rgbif package (Chamberlain &amp; Boettiger, 2017). Occurrences requested were living specimens only, having been identified after 1960, with coordinate uncertainty less than 10 kilometers, and with a location within the Northeast America region. We further filtered occurrences to ensure accuracy, including removing those with geospatial issues, coordinates in country or continent centroids, coordinates in bodies of water, and/or coordinates with a value of 0. All occurrence filtering steps were completed using detailed, custom download requests to the GBIF API.&nbsp;</p>
<p><em>Phenology</em></p>
<p>We collected phenology data for each species from iNaturalist using the rinat package in R (Barve &amp; Hart, 2022). For each species, all observations annotated as having the phenological state of “flowering” were downloaded. Phenological observations ranged in observation year from 2008 to present and are within the extent of our analyses in Northeast America. For our study, we defined Northeast America as the region bounded by 38 and 48 degrees latitude and -97.5 and -62.5 degrees longitude. This region minimizes latitudinal range to reduce inaccuracy associated with latitudinal gradients in phenological timing, while allowing for communities from a range of ecosystems.</p>
<p>From these observations, we gathered the range of months each species was observed in bloom. We removed any months that made up less than 12.5% of the total observations for a species, as these were determined probable outliers or erroneous observations. The 12.5% restriction led to data that most closely matched phenological information from online resources and additional field guides. Methods and best practices for using citizen science phenology data are still in development, so the manual identification of a best threshold given our data was deemed necessary (Barve et al., 2020). Species for which phenological data was not available from iNaturalist were removed from all analyses involving phenology.</p>
<p><em>Communities</em></p>
<p>Using phenological data, we grouped species by the months they bloom in our study area. Many species are included in multiple month groups, indicating a species that blooms across multiple months in our study region. For each month, a 10 km2 cell area . raster was generated for all species in bloom in each cell using the speciesRasterFromOccurrences() function in the speciesRaster package (Title, 2017). Presence/absence matrices were generated from each raster for each species, creating a presence/absence matrix of all species flowering during a given month in each 10 km2 community across our study area. A community, for the sake of our analyses, consists of one site during one month. Due to computational constraints and the nature of our site selection (detailed later), analyses were restricted to months for which greater than or equal to 100 species were present in our data across communities, which left 6 months included in our study: April, May, June, July, August, and September.&nbsp;</p>
<p><em>Site Selection</em></p>
<p>Due to computational constraints, our analyses were run on 100 well-sampled sites across Northeastern America. We selected sites at random from the above described set of sites containing at least 100 species recorded in our peak flowering month, June. Analyses were performed on a total of 600 communities, spanning the same 100 sites across 6 months.&nbsp;</p>
<p><img src="images/figure.png" class="img-fluid"></p>
<p><u><em>Figure 2</em></u><em>: a) Our occurrence data was restricted to the area of Northeastern America which Newcomb’s Wildflower Guide most closely describes. Red circles indicate the location of one of our 100 selected 10km2 study sites. b) Each site consists of six communities, one for each month used in our analyses. A community is made up of only the species present at that site and in bloom during the specified month. Our data, including the distribution of floral color and phylogenetic dispersion, is calculated for each of 6 months across 100 sites. Colors on pie charts represent the corresponding floral color category, with sections representing the proportion of the community having that floral color. Low phylogenetic dispersion values indicate phylogenetic underdispersion, where species are more closely related than a randomly assembled community of equal species richness; high phylogenetic dispersion values indicate phylogenetic overdispersion, where species are more distantly related than a randomly assembled community of equal species richness.</em></p>
<p><em>Floral Color Diversity Analysis</em></p>
<p>Our wildflower species span nine categories of floral color, with species in at least one, but often multiple, categories. To analyze whether a community tended towards more similar (underdispersed) or dissimilar (overdispersed) floral colors, we considered the evenness of our community across color categories. For communities, we summed the number of flowers in each color category to determine the local distribution of floral colors. We then calculated the evenness of this distribution to determine how floral colors are distributed across available categories. By calculating evenness on raw category membership, rather than a functional trait analysis on a species-by-species basis, we control for the effect of species richness, and isolate only the distribution of a community of flowers throughout color space.</p>
<p>We calculated evenness using Simpson’s Evenness (SE) via the function simpson_e from the abdiv package in R (Bittinger, 2020). SE is calculated as the inverse Simpson index (1/D) divided by the number of occurrences (in this case, the number of colors in the community) – that is, 1/DS. SE ranges from values of 0 to 1. Lower SE values indicate less even distributions, in which flowers are more grouped into particular color categories, consistent with convergent (underdispersed) floral colors. Higher SE values indicate more even distributions, in which flowers in a community are spread across most or all color categories, consistent with divergent (overdispersed) floral colors. We selected SE as our metric for its robustness with respect to uneven distributions of flowers across color categories in the overall dataset. Across all species, the frequency of common colors, e.g.&nbsp;white, are greater by an order of magnitude than the frequency of rarer colors, e.g.&nbsp;orange. As such, an appropriate evenness metric must primarily measure the evenness of distribution across common color categories, without being heavily influenced by rarer categories (Smith &amp; Willson, 1996). To test whether flower color was more or less even than expected at random across communities, we conducted a randomization test. For each community, we generated a null distribution of evenness values by holding species richness constant but selecting random species from our overall species pool. This approach accounts for any artifacts in our data driven solely by differences in species richness across our communities. We repeated this randomization 100 times in each community to generate permutation tests. We tested the difference between our observed and random communities using an ANOVA test and Welch two-sample t-test to determine the significance of the difference between the means of each distribution.&nbsp;</p>
<p><img src="images/dist1.png" class="img-fluid"></p>
<p><img src="images/dist 2.png" class="img-fluid"></p>
<p><u><em>Figure 3</em></u><em>: a) In a community with lower evenness – more uneven with respect to floral color – distribution of species across color categories is dominated by a few key colors. This is consistent with patterns of convergent floral color, where species in a community tend towards particular colors. b) In a community with higher evenness – more even with respect to floral color – distribution of species across color categories is relatively even. This is consistent with patterns of divergent floral color, where flowers within a community tend to be distinct from those they co-occur with.</em></p>
<p><em>Phylogenetic Diversity Metrics</em></p>
<p>All species were placed on a phylogeny using the V.PhyloMaker2 package (Jin &amp; Qian, 2022). The phylogeny was created according to the Leipzig Catalogue of Vascular Plants (LCVP) database using the GBOTB.extended.LCVP.tre megatree. This megatree represents each genus as a polytomy.&nbsp;</p>
<p>We calculated phylogenetic dispersion using standardized effect size of phylogenetic diversity (PD), mean pairwise distances (MPD), and mean nearest taxon distances (MNTD) for each of our 600 communities using the picante package in R (Kembel et al., 2010). These metrics offer insight into which communities are more or less phylogenetically diverse than expected from a random draw of species. Each metric offers a measure of the phylogenetic dispersion of the community, with values greater than 0 indicating an overdispersed community, and values less than 0 indicating an underdispersed community.</p>
<p>We tested for correlations between community evenness of our sites and both community species richness and community phylogenetic dispersion (PD, MPD, and MNTD) using linear regression models.</p>
</section>
<section id="results" class="level2">
<h2 class="anchored" data-anchor-id="results">Results</h2>
<p><em>Data Overview</em></p>
<p>We collected data and assembled a phylogeny for 1,096 wildflower species, spanning 473 genera from 107 families. Of our genera, 268 are represented by a single species. Our most-represented genera include Viola (n=30), Solidago (n=26), and Platanthera (n=17). The most common color categories in our phylogeny are white (n=416) and yellow (n=308), with all other colors being represented by fewer than 300 species. Our least represented color categories were brown (n=15) and orange (n=20).&nbsp;</p>
<p>From iNaturalist, we acquired phenology data for 1,019 of our species. Not surprisingly, the most frequent months for our species to be flowering were throughout late spring and summer, including July (n=595), June (n=530), and August (n=480). Of the species with phenological data available, we acquired occurrence data from GBIF for 961. The number of species in each community was roughly normally distributed (mean=38.46, sd=19.6, min=11, max=102).&nbsp;</p>
<p>Phylogenetic dispersion of each community was measured using the standard effect sizes of phylogenetic diversity (PD), mean pairwise distances (MPD), and mean nearest taxon distances (MNTD). All measures of PD were roughly normally distributed. The standard effect size of phylogenetic diversity had a mean of -0.315 (sd=1.069, min=-3.454, max=2.372). The standard effect size of MPD had a mean of 0.308 (sd=0.97, min=-2.62, max=2.04). The standard effect size of MNTD had a mean of -0.21 (sd=1.015, min=-3.61, max=2.935). Communities ranged greatly from phylogenetically under- to overdispersed according to all metrics, but were normally distributed with means all within +/-0.32 of 0.&nbsp;</p>
<p><em>Patterns of Floral Color</em></p>
<p>Simpson’s Evenness for floral color categories across communities was roughly normally distributed (mean=0.679, sd=0.079, min=0.434, max=1). The mean evenness of our observed communities was significantly lower than the mean evenness of our randomized communities (mean=0.692, sd=0.09, min=0.363, max=1), as calculated by ANOVA (df=1, F=12.86, p=0.000366) and a Welch two-sample t-test (df=613.19, t=-4.0196, p=0.00007).</p>
<p><img src="images/Rplot02.png" class="img-fluid"></p>
<p><u><em>Figure 4</em></u><em>: The distribution of observed color evenness in each of our 600 communities (in red), plotted alongside the distribution of our random assemblages, controlling for species richness, consisting of 60,000 simulated communities (in blue). The mean evenness of our observed data is significantly lower than the mean evenness of our null distribution for each community (p=0.000366).</em></p>
<p><em>Incorporating Species Relationships</em></p>
<p>Using linear regression models, we found no significant relationship between phylogenetic dispersion of communities and their color evenness using any of our phylogenetic dispersion metrics (PD, MPD, MNTD). However, color evenness of communities was significantly negatively correlated with species richness (t=-10.02, p &lt; 2e-16), emphasizing the importance of accounting for species richness in our randomization test.</p>
<p><img src="images/relationships.png" class="img-fluid"></p>
<p><u><em>Figure 5:</em></u> <em>a) No correlation was identified between the evenness of color in our communities and any of our measures of phylogenetic dispersion, including phylogenetic diversity (PD), mean pairwise distance (MPD), or mean nearest taxon distance (MNTD). This indicates that evenness of distribution across color categories in a community is not driven by phylogenetic relationships within that community – surprisingly, this means that communities with many closely related species (low phylogenetic dispersion) do not tend to converge upon particular colors any more than other communities. b) A significant negative correlation was identified between evenness of color in our communities and the species richness of those communities. This is surprising, as it indicates flowers do not tend to fill less-occupied color categories, but rather, converge towards more dominant color categories as species richness increases.</em></p>
</section>
<section id="discussion" class="level2">
<h2 class="anchored" data-anchor-id="discussion">Discussion</h2>
<p>Floral color is a complex trait, whose development and diversification is hypothesized to be driven by a wide range of biotic and abiotic factors. Understanding how floral color is structured in communities can provide crucial starting points for understanding the ecological drivers of color. Here, we examined the emergent patterns of floral color across common North American Eastern Wildflower communities, testing for patterns consistent with over- and under-dispersion of color relative to random patterns. We found that, across 600 10km2 communities encompassing six months of flowering and 1,096 species, floral color was significantly less evenly distributed than expected from random assemblage of species in communities. However, the phylogenetic dispersion of communities had no significant correlation with the evenness of floral color in that community. This indicates that emergent patterns of floral color diversity at the broad scale are not predictably driven by the phylogenetic relationships of species within a community. These results shed light on the drivers of floral color in communities at this scale, and suggest exciting avenues for future research to understand floral color diversity.</p>
<p><em>Patterns of Floral Color</em></p>
<p>We studied 600 communities, spanning the assemblage of flowering species for 6 different months across 100 sites in Northeastern America. By evaluating the evenness of distribution of flowering species across broad color categories, we sought to identify whether communities were more or less diverse in floral color than expected by chance. We found that observed floral color evenness was significantly lower than expected by chance across communities. This result indicates that, in our study region, co-occurring wildflowers tend towards more similar colors than predicted by random assemblages. This pattern is consistent with hypothesized processes producing convergent floral color within communities. Appealing to the dominant pollinator in a given season, as well as appealing to pollinators with existing floral color preferences (pollinator constancy), may promote the co-occurrence of similar floral colors (Kemp et al., 2019; Bergamo et al., 2020; Erickson &amp; Pessoa, 2021). Our findings suggest that convergent processes are outweighing the effects of any processes that promote divergence or overdispersion of floral colors (e.g., to reduce competition for pollinators and avoid deleterious hybridizations, Grossenbacher &amp; Whittall, 2014; Erickson &amp; Pessoa, 2022). Abiotic conditions, including climate and soil, may also explain how wildflower communities overcome competing biotic pressures to display convergent floral colors. The impacts of abiotic factors, such as climate and soil, on floral color is an open area of research with much work to do; further research into this area will help elucidate the role of these pressures in generating emergent patterns of floral color across landscapes (Willmer, 2011). In particular, experiments testing for evidence of similar flower colors impacting plant success in communities will be necessary to link these broad scale patterns with hypothesized processes.</p>
<p>We also identified a negative relationship between floral color evenness and species richness across communities. This negative pattern is surprising, as evenness might be expected to increase with increasing species richness; this pattern is in disagreement with Fox’s Assembly Rule, which suggests that emptier categories should be filled before fuller ones in a community (Fox &amp; Brown 1993). This pattern is consistent with a small number of floral color categories becoming increasingly dominant as species richness increases. Fortunately, our permutation-style tests against a null hypothesis of random assemblage are robust with respect to species richness. Future research may seek to disentangle which, if any, floral colors exhibit dominance in communities, and how floral color is assembled with respect to species richness. This result also lends itself to new hypotheses to explore – for instance, are emergent patterns of floral color driven in part by species richness at a site? Is it more beneficial to have a rare floral color in communities with fewer species, while in communities with more species it is more beneficial to display the locally dominant floral color?</p>
<p><em>Incorporating Species Relationships</em></p>
<p>Interestingly, when accounting for phylogenetic relationships within our communities, we saw no significant correlation between a communities’ evenness of distribution across color categories and its phylogenetic over- or underdispersion. Our study included multiple communities showing notable phylogenetic over- and underdispersion, yet the floral color distribution in these sites was no different than communities with less notable phylogenetic relationships. This suggests that phylogenetically structured processes are not a key driver of the floral color similarity patterns observed across communities at this scale.</p>
<p>Our study did not explicitly address flower color evolutionary rates across clades. However, future work with higher phylogenetic resolution could incorporate information on the degree of phylogenetic signal in flower color across subclades. For example, if floral color is phylogenetically conserved, we may expect to see significantly less even floral color distributions in phylogenetically underdispersed communities, as the close relatives in these would be expected to occupy similar color categories. The resolution of our phylogeny, with polytomies for all congeners, limits our ability to make detailed estimates of phylogenetic patterns, and may mask correlations between phylogenetic relationships and color evenness. Future research into floral color and phylogenetic patterns across communities using more detailed data, including higher-resolution phylogenies are a clear next step in this area.</p>
<p><em>Study Limitations &amp; Future Directions</em></p>
<p>Our study is, to our knowledge, the first of its kind to assemble such a large dataset of color in Northeastern American wildflowers. Due to the number of species involved in our study, many aspects of our data have room for further exploration. Our phenology, phylogeny, and color data could all be improved by approaches to using citizen science, including the creation of novel methods to acquire wildflower color data from citizen science imagery (Barve et al., 2020). By combining our methods with higher quality data on wildflower traits, we may be able to further strengthen our understanding of the drivers of floral color diversity.</p>
<p>Further, while the average evenness of color in our wildflower communities is significantly lower than expected by chance, additional analyses and more data may reveal other emergent patterns of floral color. We were limited by computational power, but future research should encompass more than 100 spatial communities across the north east. Additionally it will be important to account for spatial and temporal non-independence of sites in future work. Such research may uncover other patterns, including convergent mechanisms discussed above, but also including random ones such as drift, and divergent ones such as competition for pollinators. Regardless, this study represents a powerful first step in investigating the emergent patterns of flower color across communities at the broad scale.</p>
</section>
<section id="acknowledgements" class="level2">
<h2 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h2>
<p>The authors thank Nate Sanders for helpful insight on methods and analyses; Matthew Hack for guidance on spatial and phylogenetic methods; the Weber Lab for reviews of manuscripts and discussions on methods; and the University of Michigan Undergraduate Research Opportunity Program for continuous support and funding over the course of the study.</p>
</section>
<section id="author-contributions" class="level2">
<h2 class="anchored" data-anchor-id="author-contributions">Author Contributions</h2>
<p>CAT and MGW conceived of and designed the study. CAT collected and analyzed the data and wrote the manuscript draft. CAT and MGW interpreted the results and revised the manuscript.</p>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<p>Ahmad, S., Chen, J., Chen, G.-Z., Huang, J., Zhou, Y., Kai, Z., Lan, S., Liu, Z.-J., &amp; Peng, D. (2022). Why Black Flowers? An Extreme Environment and Molecular Perspective of Black Color Accumulation in the Ornamental and Food Crops. Frontiers in Plant Science, 13, 885176.<a href="https://doi.org/10.3389/fpls.2022.885176" class="uri">https://doi.org/10.3389/fpls.2022.885176</a></p>
<p>Barve, V., &amp; Hart, E. (2022). rinat: Access “iNaturalist” Data Through APIs.<a href="https://cran.r-project.org/package=rinat">https://CRAN.R-project.org/package=rinat</a></p>
<p>Barve, V. V., Brenskelle, L., Li, D., Stucky, B. J., Barve, N. V., Hantak, M. M., McLean, B. S., Paluh, D. J., Oswald, J. A., Belitz, M. W., Folk, R. A., &amp; Guralnick, R. P. (2020). Methods for broad-scale plant phenology assessments using citizen scientists’ photographs. Applications in Plant Sciences, 8(1), e11315.<a href="https://doi.org/10.1002/aps3.11315" class="uri">https://doi.org/10.1002/aps3.11315</a></p>
<p>Bergamo, P. J., Streher, N. S., Wolowski, M., &amp; Sazima, M. (2020). Pollinator-mediated facilitation is associated with floral abundance, trait similarity and enhanced community-level fitness. Journal of Ecology, 108(4), 1334–1346.<a href="https://doi.org/10.1111/1365-2745.13348" class="uri">https://doi.org/10.1111/1365-2745.13348</a></p>
<p>Bittinger, K. (2020). abdiv: Alpha and Beta Diversity Measures.<a href="https://cran.r-project.org/package=abdiv">https://CRAN.R-project.org/package=abdiv</a></p>
<p>Chamberlain, S., &amp; Boettiger, C. (2017). R Python, and Ruby clients for GBIF species occurrence data. PeerJ PrePrints.<a href="https://doi.org/10.7287/peerj.preprints.3304v1" class="uri">https://doi.org/10.7287/peerj.preprints.3304v1</a></p>
<p>Clegg, M. T., &amp; Durbin, M. L. (2000). Flower color variation: A model for the experimental study of evolution. Proceedings of the National Academy of Sciences, 97(13), 7016–7023.<a href="https://doi.org/10.1073/pnas.97.13.7016" class="uri">https://doi.org/10.1073/pnas.97.13.7016</a></p>
<p>Erickson, M. F., &amp; Pessoa, D. M. A. (2022). Determining factors of flower coloration. Acta Botanica Brasilica, 36.<a href="https://doi.org/10.1590/0102-33062021abb0299" class="uri">https://doi.org/10.1590/0102-33062021abb0299</a></p>
<p>Fox, B. J., &amp; Brown, J. H. (1993). Assembly Rules for Functional Groups in North American Desert Rodent Communities. Oikos, 67(2), 358–370.<a href="https://doi.org/10.2307/3545483" class="uri">https://doi.org/10.2307/3545483</a></p>
<p>Freiberg, M., Winter, M., Gentile, A., Zizka, A., Muellner-Riehl, A. N., Weigelt, A., &amp; Wirth, C. (2020). LCVP, The Leipzig catalogue of vascular plants, a new taxonomic reference list for all known vascular plants. bioRxiv, 0, 0.<a href="https://doi.org/10.1101/2020.05.08.077149" class="uri">https://doi.org/10.1101/2020.05.08.077149</a></p>
<p>Grossenbacher, D. L., &amp; Whittall, J. B. (2011). INCREASED FLORAL DIVERGENCE IN SYMPATRIC MONKEYFLOWERS. Evolution, 65(9), 2712–2718.<a href="https://doi.org/10.1111/j.1558-5646.2011.01306.x" class="uri">https://doi.org/10.1111/j.1558-5646.2011.01306.x</a></p>
<p>Jin, Y., &amp; Qian, H. (2022). V.PhyloMaker2: An updated and enlarged R package that can generate very large phylogenies for vascular plants. Plant Diversity, 44(4), 335–339.<a href="https://doi.org/10.1016/j.pld.2022.05.005" class="uri">https://doi.org/10.1016/j.pld.2022.05.005</a></p>
<p>Kembel, S. W., Cowan, P. D., Helmus, M. R., Cornwell, W. K., Morlon, H., Ackerly, D. D., Blomberg, S. P., &amp; Webb, C. O. (2010). Picante: R tools for integrating phylogenies and ecology. Bioinformatics, 26, 1463–1464.</p>
<p>Kemp, J. E., Bergh, N. G., Soares, M., &amp; Ellis, A. G. (2019). Dominant pollinators drive non-random community assembly and shared flower colour patterns in daisy communities. Annals of Botany, 123(2), 277–288.<a href="https://doi.org/10.1093/aob/mcy126" class="uri">https://doi.org/10.1093/aob/mcy126</a></p>
<p>Koski, M. H., &amp; Ashman, T.-L. (2016). Macroevolutionary patterns of ultraviolet floral pigmentation explained by geography and associated bioclimatic factors. New Phytologist, 211(2), 708–718.<a href="https://doi.org/10.1111/nph.13921" class="uri">https://doi.org/10.1111/nph.13921</a></p>
<p>Muñoz-Galicia, D., Castillo-Guevara, C., &amp; Lara, C. (2021). Innate and learnt color preferences in the common green-eyed white butterfly (Leptophobia aripa): Experimental evidence. PeerJ, 9, e12567.<a href="https://doi.org/10.7717/peerj.12567" class="uri">https://doi.org/10.7717/peerj.12567</a></p>
<p>Newcomb, L. (1977). Newcomb’s Wildflower Guide. Little, Brown and Company.</p>
<p>Pebesma, E. (2018). Simple Features for R: Standardized Support for Spatial Vector Data. The R Journal, 10(1), 439–446.<a href="https://doi.org/10.32614/RJ-2018-009" class="uri">https://doi.org/10.32614/RJ-2018-009</a></p>
<p>Rabosky, A. D., Cox, C., Rabosky, D., Title, P., Holmes, I., Feldman, A., &amp; McGuire, J. (2016). Coral snakes predict the evolution of mimicry across New World snakes. Nature Communications, 7, 11484.<a href="https://doi.org/10.1038/ncomms11484" class="uri">https://doi.org/10.1038/ncomms11484</a></p>
<p>Reverté, S., Retana, J., Gómez, J. M., &amp; Bosch, J. (2016). Pollinators show flower colour preferences but flowers with similar colours do not attract similar pollinators. Annals of Botany, 118(2), 249–257.<a href="https://doi.org/10.1093/aob/mcw103" class="uri">https://doi.org/10.1093/aob/mcw103</a></p>
<p>Sapir, Y., Gallagher, M. K., &amp; Senden, E. (2021). What Maintains Flower Colour Variation within Populations? Trends in Ecology &amp; Evolution, 36(6), 507–519.<a href="https://doi.org/10.1016/j.tree.2021.01.011" class="uri">https://doi.org/10.1016/j.tree.2021.01.011</a></p>
<p>Smith, B., &amp; Wilson, J. B. (1996). A Consumer’s Guide to Evenness Indices. Oikos, 76(1), 70–82.<a href="https://doi.org/10.2307/3545749" class="uri">https://doi.org/10.2307/3545749</a></p>
<p>Sobel, J. M., &amp; Streisfeld, M. A. (2013). Flower color as a model system for studies of plant evo-devo. Frontiers in Plant Science, 4.<a href="https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2013.00321" class="uri">https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2013.00321</a></p>
<p>Title, P. (2017). speciesRaster: Generation of SpeciesRaster Object and Calculation of Morphological and Phylogenetic Metrics.</p>
<p>Trunschke, J., Lunau, K., Pyke, G. H., Ren, Z.-X., &amp; Wang, H. (2021). Flower Color Evolution and the Evidence of Pollinator-Mediated Selection. Frontiers in Plant Science, 12.<a href="https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2021.617851" class="uri">https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2021.617851</a></p>
<p>Warren, J., &amp; Mackenzie, S. (2001). Why are all colour combinations not equally represented as flower-colour polymorphisms? New Phytologist, 151(1), 237–241.<a href="https://doi.org/10.1046/j.1469-8137.2001.00159.x" class="uri">https://doi.org/10.1046/j.1469-8137.2001.00159.x</a></p>
<p>Weiss, M. R., &amp; Papaj, D. R. (2003). Colour learning in two behavioural contexts: How much can a butterfly keep in mind? Animal Behaviour, 65(3), 425–434.<a href="https://doi.org/10.1006/anbe.2003.2084" class="uri">https://doi.org/10.1006/anbe.2003.2084</a></p>
<p>Willmer, P. (2011). Pollination and Floral Ecology. In Pollination and Floral Ecology. Princeton University Press.<a href="https://doi.org/10.1515/9781400838943" class="uri">https://doi.org/10.1515/9781400838943</a></p>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Copyright 2024, Chris Talbot</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/chris-a-talbot">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/chris__talbot">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/chris-talbot">
      <i class="bi bi-linkedin" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>